proposals/NNNN-android-transport-overhaul.md

```diff
@@ -0,0 +1,490 @@
+			this.priorityForBuffingMessage = receivedBundle.getInt(TransportConstants.PACKET_PRIORITY_COEFFICIENT,0);
+		}
+		handleMessage(flags, packet, transportType);
+	}else{
+		//Add the write task on the stack
+		PacketWriteTaskBlockingQueue queue = queues.get(transportType);
+		if(queue == null){	//TODO check to see if there is any better place to put this
+			queue = new PacketWriteTaskBlockingQueue();
+			queues.put(transportType,queue);
+		}
+		if(queue!=null){
+			queue.add(new PacketWriteTask(receivedBundle));
+			PacketWriteTaskMaster packetWriteTaskMaster = packetWriteTaskMasterMap.get(transportType);
+			if(packetWriteTaskMaster!=null){
+				packetWriteTaskMaster.alert();
+			}
+		}
+	}
+	return true;
+}
+
+```
+
+#### SdlPacket Update
+
+The `SdlPacket` class will now include a `TransportRecord` for the local library. This lets the router service know which transport to send a packet. Also, when the router service receives a packet from the core module and relays it to the bound client app, the client knows where the packet came from.
+
+```java
+public class TransportRecord implements Parcelable{
+
+    TransportType type;
+    String address;
+
+    public TransportRecord(TransportType transportType, String address){
+        this.type = transportType;
+        this.address = address;
+    }
+
+    ....    
+}
+```
+
+
+#### Secondary Transport Registration
+
+If the router service notices a Register Transport control frame ACK is received it will look for the corresponding session, register that transport for that session on that registered app, and forward the packet to the client.
+
+### Breaking change phase
+
+
+#### Remove WiProProtocol, SdlConnection, and Listener
+
+All of these will no longer be necessary and should be removed.
+
+#### Move SdlSession Class
+
+With `SdlConnection` and listener removed, the package `com.smartdevicelink.SdlConnection` will no longer be valid and keeping `SdlSession` there will be orphaned. The class will be moved to `com.smartdevicelink.proxy`.
+
+#### Misc Cleanup of Deprecated items
+
+It will be required to remove all deprecated methods previously by this proposal's implementation.
+
+## Potential downsides
+
+- All transports will have to rely on the router service
+- For multiple sessions per app the initial design requires an app to create a new proxy and hold a separate binding to the router service. This can be overcome in the future but as the use case is very small it seems not worth the effort. 
+
+## Impact on existing code
+
+- New classes would be created to forge a much more straightforward transport layer.
+- `SdlConnection.java` would be deprecated.
+- Deprecate all transport configs in favor of a new general one.
+
+## Alternatives considered
+- Continuing with the current stack and implemented the multiple transports and USB transport to the router service. This method turns out to be very messy and changing in the future becomes difficult. 
```

Comment:
I think this should read "...and implementing the multiple transports and USB transport into the router service."